src/main/java/com/couchbase/client/commons/CouchbaseMap.java

```diff
@@ -0,0 +1,310 @@
+ * respectively.
+ *
+ * @param <V> the type of values in the map (restricted to {@link JsonObject}.
+ */
+public class CouchbaseMap<V> extends AbstractMap<String, V> {
+
+    public static final int MAX_OPTIMISTIC_LOCKING_ATTEMPTS = CouchbaseArrayList.MAX_OPTIMISTIC_LOCKING_ATTEMPTS;
+
+    private final String id;
+    private final Bucket bucket;
+
+    /**
+     * Create a new {@link CouchbaseMap}, backed by the document identified by <code>id</code>
+     * in the given Couchbase <code>bucket</code>. Note that if the document already exists,
+     * its content will be used as initial content for this collection. Otherwise it is created empty.
+     *
+     * @param id the id of the Couchbase document to back the map.
+     * @param bucket the {@link Bucket} through which to interact with the document.
+     */
+    public CouchbaseMap(String id, Bucket bucket) {
+        this.id = id;
+        this.bucket = bucket;
+
+        try {
+            bucket.insert(JsonDocument.create(id, JsonObject.empty()));
+        } catch (DocumentAlreadyExistsException ex) {
+            // Ignore concurrent creations, keep on moving.
+        }
+    }
+
+    /**
+     * Create a new {@link CouchbaseMap}, backed by the document identified by <code>id</code>
+     * in the given Couchbase <code>bucket</code>. Note that if the document already exists, its content is
+     * reset to the values copied from the given <code>data</code> Map.
+     *
+     * A null or empty data map will re-initialize any pre-existing document to an empty content.
+     *
+     * @param id the id of the Couchbase document to back the list.
+     * @param bucket the {@link Bucket} through which to interact with the document.
+     * @param data Map of the elements to initially store in the CouchbaseMap.
+     */
+    public CouchbaseMap(String id, Bucket bucket, Map<String, ? extends V> data) {
+        this.id = id;
+        this.bucket = bucket;
+
+        JsonObject content = JsonObject.create();
+        if (data != null && !data.isEmpty()) {
+            for (Entry<String, ? extends V> entry : data.entrySet()) {
+                if (entry.getKey() == null) {
+                    throw new NullPointerException("Attempted to create a map with a null key");
+                }
+                content.put(entry.getKey(), entry.getValue());
+            }
+        }
+        JsonDocument initial = JsonDocument.create(id, content);
+        bucket.upsert(initial);
+    }
+
+    @Override
+    public V put(String key, V value) {
+        if (key == null) {
+            throw new NullPointerException("Unsupported null key");
+        }
+        if (!JsonValue.checkType(value)) {
+            throw new IllegalArgumentException("Unsupported value type.");
+        }
+
+        for(int i = 0; i < MAX_OPTIMISTIC_LOCKING_ATTEMPTS; i++) {
+            try {
+                DocumentFragment<Lookup> current = bucket.lookupIn(id).get(key).execute();
```

Comment:
no need, for `put` the `Map` interface uses the `K` type (here `String`) directly, unlike in the other methods.